cli/file_fetcher.rs

```diff
@@ -2,677 +2,499 @@
+        loop {
+          if list.contains(&path.to_str().unwrap().to_string()) {
+            return false;
+          }
+          if !path.pop() {
+            break;
+          }
+        }
+        true
+      }
     }
-
-    Ok(())
   }
+}
 
-  /// Required for TS compiler and source maps.
-  pub fn fetch_cached_source_file(
-    &self,
-    specifier: &ModuleSpecifier,
-    permissions: Permissions,
-  ) -> Option<SourceFile> {
-    let maybe_source_file = self.source_file_cache.get(specifier.to_string());
-
-    if maybe_source_file.is_some() {
-      return maybe_source_file;
-    }
-
-    // If file is not in memory cache check if it can be found
-    // in local cache - which effectively means trying to fetch
-    // using "--cached-only" flag.
-    // It should be safe to for caller block on this
-    // future, because it doesn't actually do any asynchronous
-    // action in that path.
-    if let Ok(maybe_source_file) =
-      self.get_source_file_from_local_cache(specifier.as_url(), &permissions)
-    {
-      return maybe_source_file;
-    }
+/// Fetch a source file from the local file system.
+fn fetch_local(specifier: &ModuleSpecifier) -> Result<File, AnyError> {
+  let local = specifier.as_url().to_file_path().map_err(|_| {
+    uri_error(format!("Invalid file path.\n  Specifier: {}", specifier))
+  })?;
+  let bytes = fs::read(local.clone())?;
+  let charset = text_encoding::detect_charset(&bytes).to_string();
+  let source = get_source_from_bytes(bytes, Some(charset))?;
+  let media_type = MediaType::from(specifier);
+
+  Ok(File {
+    local,
+    maybe_types: None,
+    media_type,
+    source,
+    specifier: specifier.clone(),
+  })
+}
 
-    None
+/// Return a validated scheme for a given module specifier.
+fn get_scheme(specifier: &ModuleSpecifier) -> Result<String, AnyError> {
+  let scheme = specifier.as_url().scheme();
+  if !SUPPORTED_SCHEMES.contains(&scheme) {
+    Err(generic_error(format!(
+      "Unsupported scheme \"{}\" for module \"{}\". Supported schemes: {:#?}",
+      scheme, specifier, SUPPORTED_SCHEMES
+    )))
+  } else {
+    Ok(scheme.to_string())
   }
+}
 
-  /// Save a given source file into cache.
-  /// Allows injection of files that normally would not present
-  /// in filesystem.
-  /// This is useful when e.g. TS compiler retrieves a custom_error file
-  /// under a dummy specifier.
-  pub fn save_source_file_in_cache(
-    &self,
-    specifier: &ModuleSpecifier,
-    file: SourceFile,
-  ) {
-    self.source_file_cache.set(specifier.to_string(), file);
+/// Given a vector of bytes and optionally a charset, decode the bytes to a
+/// string.
+fn get_source_from_bytes(
+  bytes: Vec<u8>,
+  maybe_charset: Option<String>,
+) -> Result<String, AnyError> {
+  let source = if let Some(charset) = maybe_charset {
+    text_encoding::convert_to_utf8(&bytes, &charset)?.to_string()
+  } else {
+    String::from_utf8(bytes)?
+  };
+  if source.starts_with("#!") {
```

Comment:
It seems odd that a function called `get_source_from_bytes` would also handle shebang stuff.

Reply:
Makes more sense to break it into two.